synapse/util/state_compressor.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import TYPE_CHECKING
+
+from synapse.logging.context import defer_to_thread
+from synapse.metrics.background_process_metrics import run_as_background_process
+
+try:
+    import auto_compressor as state_compressor
+except ImportError:
+    state_compressor = None
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+
+def setup_state_compressor(hs: "HomeServer"):
+    """Schedules the state compressor to run regularly"""
+
+    # Return if cannot import auto_compressor
+    if not state_compressor or not hs.config.workers.run_background_tasks:
+        return
+
+    # Return if compressor isn't enabled
+    compressor_config = hs.config.statecompressor
+    if not compressor_config.compressor_enabled:
+        return
+
+    # Check that the database being used is postgres
+    db_config = hs.config.database.get_single_database().config
+    if db_config["name"] != "psycopg2":
+        return
+
+    # construct the database URL from the database config
+    db_args = db_config["args"]
+    db_url = "postgresql://{username}:{password}@{host}:{port}/{database}".format(
+        username=db_args["user"],
+        password=db_args["password"],
+        host=db_args["host"],
+        port=db_args["port"],
+        database=db_args["database"],
```

Comment:
Bleurgh, a bunch of these are actually optional values. Annoyingly host is optional in Synapse (which uses psycopg2 which uses libpq) but not in Rust, and its Hard to know what the default host is in libpq without asking it. If we want this to work I think we need to explicitly pull out `db_conn.info.host` of a real connection to the state database.

Reply:
I think this now works with only password still being taken from db_args (since psycopg2 doesn't provide that in it's dsn_parameters field - see [here](https://www.psycopg.org/docs/extensions.html#psycopg2.extensions.ConnectionInfo.dsn_parameters))